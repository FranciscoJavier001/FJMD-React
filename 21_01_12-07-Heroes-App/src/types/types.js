//**_______________________________________________________________________________________________________________________________________________*/
//** Este objeto lo voy a exportar, que contiene los types de la aplicacion */

//** Asi voy a manejar los types */
export const types = {
    login: '[auth] login', //** Se dispara cuando se hace el login */
    logout: '[auth] logout' //** Se ddispara cuando se hace el logout */
}