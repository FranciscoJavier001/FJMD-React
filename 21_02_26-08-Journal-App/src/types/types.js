

export const types = {
    //** Aqui voy a definir los tipos */

    login: '[Auth] Login', //** Este apunta al Auth Login, */
    logout: '[Auth] Logout', //** Este apunta al Auth LogOut, */

    uiSetError: '[UI] Set Error',
    uiRemoveError: '[UI] Remove Error'
}