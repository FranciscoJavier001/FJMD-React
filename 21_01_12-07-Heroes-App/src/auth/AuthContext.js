//**_______________________________________________________________________________________________________________________________________________*/
//** Este me permite crear un contexto en toda la aplicacion - La primera mayuscula porque es un HOC */
import { createContext } from 'react'

export const AuthContext =  createContext(); //** Esta en src/HeroesApp y components/login/LoginScreen, lo utilizo con el dispatch */