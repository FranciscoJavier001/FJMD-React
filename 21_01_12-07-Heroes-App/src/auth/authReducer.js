//** Esto es una simple funcion pura */

import { types } from "../types/types";

// const state = { //** Voy a tener una estado que sea parecido a este que tenga el name del usuario y el logged en true, esto va a ser mi estado si el usuario esta autentificado, y si no esta autentificado va a tener el logged en falso */
//     name: 'Francisco',
//     logged: true
// }

export const authReducer = ( state = {}, action ) => { //** Esta funcion recibe el state que va a ser un objeto y recibe el action que va a ser lo que quiero hacer y esta es una funcion de flecha y la vamos a exportar para utilizar en otro lugar y dentro voy a tener un switch que vamos avaluar el action.type */

    switch ( action.type ) {
        case types.login: //** Aqui ingresamos las aciones, uno es cuando esta en el login y en el logout, importamos el types y escribimos la accion que yo quiero despues del punto */
            return { //** Para el return voy a retornar todo lo que contenga el action.payload */
                ...action.payload,
                //** Voy a poner otra propiedad interesante que va a ser la autentificacion del usuario que va a ser logged */
                logged: true //** Va a ser true si pasa la autentifficacion */
            }

            case types.logout:
                return { //** Voy a hacer el return del objeto donde no me interese nada lo que tenga en el payload, simplemente que tenga el logged en false */
                    logged: false
                }
    
        default: //** El default va a ser el state asi como se encuentre */
            return state;
    }
}