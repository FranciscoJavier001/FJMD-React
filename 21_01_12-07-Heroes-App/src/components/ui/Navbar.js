//**_______________________________________________________________________________________________________________________________________________*/
//** Esto es un Navbar - Lo copie del Git */
import React from 'react'
import { Link, NavLink } from 'react-router-dom'


//  export const Navbar = ({ history }) => { //** Aqui en pego el history, no compartir mucha informacion por los props */
export const Navbar = () => { //** Es la linea 8, pero remplazo la sintaxis */


    return (
        //** navbar=Para que todo este junto y no haya espacios */
        //** navbar-expand-sm=para poner el navbar en fila y separados */
        //** navbar-dark=letras en color gris */
        //** bg-dark=fondo en negro */
        <nav className="navbar navbar-expand-lg navbar-dark bg-dark">
            
            <Link 
                className="navbar-brand" //** Es para que salga al inicio y mas grande */
                to="/" //** Donde me lleva cuando hago click */
            >
                HeroesApp {/* Texto */}
            </Link>

            <div className="navbar-collapse col-1"> {/* Hacer mas chico la anchura */}
                <div className="navbar-nav">{/* Para que todas esten en hileras */}

                    <NavLink 
                        activeClassName="active" //** Dond este seleccionado se quede en color blanco */
                        className="nav-item nav-link" //** nav-item=Item de navegacion - nav-link= es para que todos se vean a la altura */
                        exact //** Es la direccion a la que lleva */
                        to="/marvel"
                    >
                        Marvel
                    </NavLink>

                    <NavLink 
                        activeClassName="active"
                        className="nav-item nav-link" 
                        exact
                        to="/dc"
                    >
                        DC
                    </NavLink>

                    <NavLink //** Ruta que me lleva al Search en el NavBar */
                        activeClassName="active" //** Para que se quede blanco cuando este ahi */
                        className="nav-item nav-link"
                        exact
                        to="/search"
                    >
                        Search
                    </NavLink>
                </div>
            </div>

          
        </nav>
    )
}