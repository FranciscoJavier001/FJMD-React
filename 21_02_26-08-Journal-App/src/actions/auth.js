//** Estas son acciones que vamos a definir */
import { firebase, googleAuthProvider } from '../firebase/firebase-config'
import { types } from '../types/types'

//** Esto es una tarea asincrona, osea que se va a ejecutar todo el proceso, luego se va a encontrar con el dispatch que ejecuta la accion del login que es la que se va a aplicar directamente en el store (la fuente unica de la verdad), que modifica el state y nos regresa todo lo que estamos esperando  */
export const startLoginEmailPassword = (email, password) => { //** Esta funcion va a recibir el email y el password */
    return (dispatch) => { //** Esta funcion va a regresar un callback, osea una funcion a disparar */
        //** Hay que llamar a otro dispatch cuando ya se ejecute todo, cuando ya tenga la data */
        setTimeout(() => {
            dispatch( login( 123, 'Pedro')) //** Aqui voy a hacer un dispatch del login, con el usuario */
        }, 3500);
    }
}

export const startRegisterWithEmailPasswordName = ( email, password, name ) => { //** Esta es una tarea asincrona, asi que voy a tener un callback */
    return ( dispatch ) => { //** Cuando tenga el usuario grabado en firebase voy a hacer el dispatch */
        firebase.auth().createUserWithEmailAndPassword( email, password ) //** Aqui mandamos los parametros de email y password */
        .then( async ({ user }) => { //** Esta va a ser una promesa */

            await user.updateProfile({ displayName: name }) //** que me de el name que estoy recibiendo como argumento de mi funcion y esto me regresa una promesa */

            dispatch( //** Esto es para llamar de nuevo el dispatch */
                login( user.uid, user.displayName)
            )
        })
        .catch( e => { //** Esto es para atrapar el error y mostrarlo */
            console.log(e);
        })
    }
}

export const startGoogleLogin = () => { //** Una funcion de fleca que no recibe nada, pero es una tarea sincrona voy a poner el return y el callback me va a proveer el dispatch, por lo cual cuando tenga la informacion lo voy a llamar */
    return ( dispatch ) => {
    firebase.auth().signInWithPopup( googleAuthProvider ) //** Esto va a retormarme una promesa */
        .then( ({ user }) => { //** Aqui voy a usar la desestructuracion para extraer el user */
            dispatch( //** Que cuando esto se responda voy a hacer el dispatch de la accion que ya tengo definido abajo con el login, donde voy a poder el id que viene del user.uid, y el displayName que viene del user.displayName */
                login( user.uid, user.displayName)
            )
        })
    }
}

export const login = (uid, displayName) => ({ //** Esta es la accion que voy a llamar cuando tenga el uid y el displayNameny esta accion me debe regresar un type.login */
    type: types.login,
    payload: { //** El payload va a ser un objeto que tenga un uid y el displayName */
        uid,
        displayName
    }
})