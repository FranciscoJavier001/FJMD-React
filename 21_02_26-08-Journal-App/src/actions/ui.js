//** Acciones en el UI */

import { types } from "../types/types";

export const setError = ( err ) => ({ //** De una vez voy a hacer el return del objeto y va a recibir el payload del error */
    type: types.uiSetError,
    payload: err
})

export const removeError = () => ({ //** De una vez voy a hacer el return del objeto y no va a recibir ningun objeto */
    type: types.uiRemoveError, //** Va a llamar a la funcion de remover error */
})

export const startLoading = () => ({ //** Asi llamamos las acciones, y ahora se debe hacer el dispatch de estas acciones */
    type: types.uiStartLoading,
})

export const finishLoading = () => ({
    type: types.uiFinishLoading,
})