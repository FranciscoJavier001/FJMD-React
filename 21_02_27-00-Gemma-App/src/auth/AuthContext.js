//** Recuerda que empieza con mayusculas porque es un HighOrderComponent */
//** Voy a hacer una importacion de createContext */
import { createContext } from 'react'

export const AuthContext =  createContext(); //** Voy a crear una constante que se llame AuthContext donde no le voy a mandar ningun argumento y lo voy a importar */

//** Este lo voy a colocar en HeroesApp ya que ahi empieza a desplegarse toda la aplicacion y aqui voy a poner el HighOrderComponent que va a ser el AuthContext */