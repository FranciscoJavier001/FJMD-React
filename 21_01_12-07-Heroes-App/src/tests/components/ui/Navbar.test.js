import React from 'react';
import { mount } from 'enzyme'
import { MemoryRouter, Router } from 'react-router-dom';
import '@testing-library/jest-dom'

import { AuthContext } from '../../../auth/AuthContext';
import { Navbar } from '../../../components/ui/Navbar';
import { types } from '../../../types/types';

describe('Pruebas en <Navbar />', () => {

    //** Que onda con los customhooks, donde lo primero que quiero hacer es generar un historyMook que es el que se llama a traves del router, en palabras simples se lo puedo mandar al router y el componente cuando va a utilizar el useHistory va a ver y a utilizar el history que tiene este router y puedo fingir el router en el wrapper = mount */
    const historyMook = { //** Este historyMook va a tener lo que queremos que tenga que puede tener un push que seria un jest.fn() para hacer igualaciones, y el location que va a ser un objeto vacio */
        // Aqui me mando un error, me pidio un listen
        push: jest.fn(),
        replace: jest.fn(),
        location: {},
        listen: jest.fn(),
        createHref: jest.fn(),
    }

    // Aqui hay que Proveer lo que tenia el Context, que es el user y el dispatch
    const contextValue = { //** Este va a ser la simulacion del objeto que definimos abajo, en el provider */
        dispatch: jest.fn(), //** Cuando queremos provar funciones lo mejor que podemos hacer es usar jest.fn() */
        user: {
            logged: true, //** Va a tener el logged en true porque el usuario ya debia estar autentificado */
            name: 'Carlos'
        }
    }
    
    //** Voy a crear el wrapper afuera, porque puede que pueda reutilizar los componentes */
    const wrapper = mount( //** Asi renderizamos componentes, y hay que vereficar si recibe algun argumento */
        // Hay que importar el AuthContext.Provider para que se pueda hacer la desestructuracion de user y no nos maque error
        // Otra cosa, el Context es un HighOrderComponent, osea que ya tenemos una estructura que se parece muchisimo a las pruebas que ya hemos echo antes en el PrivateRoute, asi que no me va a servir con el shallow, voy a tener que usar el mount
        // El contextValue lo voy a mandar al Provider
        <AuthContext.Provider value={ contextValue }>
            <MemoryRouter> 
                {/* Aqui fue donde fingimos el router y vamos a colocar el NavBar dentro del router, y el router recibe el history y dentro del history vamos a mandar el historyMook */}
                <Router history={ historyMook }>
                    <Navbar /> 
                </Router>
            </MemoryRouter>
        </AuthContext.Provider>
    );

    // Asi es para limpiar el mook, despues de cada una de las pruebas o antes
    afterEach(() => {
        jest.clearAllMocks();
    })

    test('Debe de mostrarse correctamente', () => {
        expect( wrapper ).toMatchSnapshot(); //** El snapshot toma una fotografia que asi lo va a mantener, para si hago algun cambio de lo que se reibio y lo que es, para que asi se quede el componente */
        expect( wrapper.find('.text-info').text().trim() ).toBe('Carlos') //** Aqui esperamos que encuentre la clase de text.info (clase de bootstrap) que tenga el texto, con el trim borramos los espacios que tenga la variable de carlos */
    });

    test('Debe de llamar el logout y el usar history', () => {
        
        //** Vamos a verificar que se llame el dispatch */
        wrapper.find('button').prop('onClick')(); //** Con el wrapper.find voy a buscar el boton y la prop que es la propiedad de la funcion del onClick, asi se llama */

        //** Aqui deberiamos evaluar si el dispatch fue llamado con que argumentos que va a ser un objeto que el tipo fue llamado con el logout , esto seria todo por la parte del dispatch*/
        expect( contextValue.dispatch ).toHaveBeenCalledWith({
            type: types.logout
        })

        //** Ahora voy a esperar que se haya llamado el replace que duera llamado con el /login */
        expect( historyMook.replace ).toHaveBeenCalledWith('/login');
    })  
})

