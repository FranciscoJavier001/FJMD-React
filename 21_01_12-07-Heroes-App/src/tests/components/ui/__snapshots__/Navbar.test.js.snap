// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <Navbar /> Debe de mostrarse correctamente 1`] = `
<nav
  className="navbar navbar-expand-lg navbar-dark bg-dark"
>
  <a
    className="navbar-brand"
    onClick={[Function]}
  >
    Asociaciones 
  </a>
  <div
    className="navbar-collapse col-1"
  >
     
    <div
      className="navbar-nav"
    >
      <a
        aria-current={null}
        className="nav-item nav-link"
        onClick={[Function]}
      >
        Marvel
      </a>
      <a
        aria-current={null}
        className="nav-item nav-link"
        onClick={[Function]}
      >
        DC
      </a>
      <a
        aria-current={null}
        className="nav-item nav-link"
        onClick={[Function]}
      >
        Search
      </a>
    </div>
  </div>
  <div
    className="navbar-collapse collapse w-100 order-3 dual-collapse2"
  >
     
    <ul
      className="navbar-nav ml-auto"
    >
       
      <span
        className="nav-item nav-link text-info"
      >
         
        Carlos
         
      </span>
      <button
        className="nav-item nav-link btn"
        onClick={[Function]}
      >
        Logout
      </button>
    </ul>
  </div>
</nav>
`;
